File: problems/c/3.c
/** Euler Project, Problem 3. 
 *Find the largest prime factor of a number.
 * 
 * Currently, this number is 600851475143.  
 **/ 

 
#include <stdio.h> 
#include <math.h> 
#include <stdlib.h> 

int prime(long n); 

int main(){
    long NUMBER = 600851475143; 
    long i; 

    for (i = sqrt(NUMBER); i > 1; i--){
        if (NUMBER%i == 0) {
            printf("i = %ld\n", i);
            if (prime(i)) break; 
        }  
    }

    printf("The largest prime factor in %ld is %ld.\n",NUMBER,i); 

    return 0;
} 

/**
 * find the number of factors for n .
 **/ 
int prime (long n){ 
    int i; 
    for (i = 2; i < sqrt(n); i++) {
        if (n%i==0) return 0; 
    } 
    return 1; 
} 

File: problems/rust/3.rs
// The prime factors of 13195 are 5, 7, 13 and 29.
// What is the largest prime factor of the number 600851475143?


fn main() { 
    let bignum = 600851475143; 
    println!("prime or not: {:b}", isPrime(num));  

    let mut i=1; 

    while i < bignum { 
        if bignum%i == 0 { 
            if isPrime(i) {
                println!("i: {:d}", i); 
            }
        }
        i += 1; 
    }
}

#[test]
fn isPrime_test_2() { 
    assert!(isPrime(2) == true); 
} 

#[test]
fn isPrime_test_3() { 
    assert!(isPrime(3) == true); 
} 

#[test]
fn isPrime_test_4() { 
    assert!(isPrime(4) == false); 
} 

fn isPrime(n: int) -> bool {
    match n { 
        2 => return true, 
        3 => return true,
        _ => { 
            if (n%2==0) || (n%3==0) {
                return false;
            } else { 
                let mut i = 5; 
                let mut w = 2; 

                while i*i < n { 
                    if n%i == 0 { 
                        return false; 
                    }
                    i += w; 
                    w = 6 - w; 
                } 
                return true; 
            } 
        }
    }
}
